src/component/handlers/edit/editOnInput.js

```diff
@@ -79,12 +80,53 @@ function editOnInput(): void {
     isBackward: false,
   });
 
-  var newContent = DraftModifier.replaceText(
-    content,
-    targetRange,
-    domText,
-    block.getInlineStyleAt(start)
-  );
+  var entityKey = block.getEntityAt(start);
+  var entity = entityKey && Entity.get(entityKey);
+  var applyEntity = true;
+  var entityType = entity && entity.getMutability();
+  var changeType = 'spellcheck-change';
+  var newContent;
+  var mergeSelection = true;
+
+  if (entityType) {
+    var isImmutableEntity = entityType === 'IMMUTABLE';
+    var isSegmentedEntity = entityType === 'SEGMENTED';
+
+    // Immutable or segmented entities cannot properly be handled by the
+    // default browser undo, so we have to use a different change type to
+    // force using our internal undo method instead of falling through to the
+    // native browser undo.
+    changeType = (isImmutableEntity || isSegmentedEntity) ?
+      'apply-entity' :
+      changeType;
+
+    if (isImmutableEntity || isSegmentedEntity) {
+      var characterRemoved = domText.length < modelText.length;
```

Comment:
@hellendag That works as expected. The case for changing a character and when it is longer is identical. Here are 3 gifs showing each behavior for each entity type:

`IMMUTABLE`
![spellcheck_immutable](https://cloud.githubusercontent.com/assets/3470207/14686927/9ab288fc-0743-11e6-8820-76e68db38730.gif)

`MUTABLE`
![spellcheck_mutable](https://cloud.githubusercontent.com/assets/3470207/14686928/9ab638da-0743-11e6-897a-adbd6e86e837.gif)

`SEGMENTED`
![spellcheck_segmented](https://cloud.githubusercontent.com/assets/3470207/14686929/9ab6d7f4-0743-11e6-99ec-a8fbdc5bc4a7.gif)